### PR TITLE
support dynamic tracing for dlopen'ed modules

### PR DESCRIPTION
**Pull request by ParkHanbum**
support instrument to modules which loaded by dynamic method dlopen.

for example : 

```
m@ubuntu:~/git/uftrace_dlopen2$ ldd `which java`
        linux-vdso.so.1 (0x00007ffcec0fc000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f86688b5000)
        libjli.so => not found
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f86686b1000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f86682c0000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f8668ad4000)

m@ubuntu:~/git/uftrace_dlopen2$ ./uftrace -L. -P.@libjvm.so java HelloWorld
Hello world!
# DURATION     TID     FUNCTION
            [110346] | JLI_Launch() {
            [110346] |   JNI_GetDefaultJavaVMInitArgs() {
   1.037 us [110346] |     Threads::is_supported_jni_version();
   4.145 us [110346] |   } /* JNI_GetDefaultJavaVMInitArgs */
            [110348] | JNI_CreateJavaVM() {
            [110348] |   Threads::create_vm() {
            [110348] |     ResourceObj::operator new() {
            [110348] |       os::malloc() {
  13.583 us [110348] |         MallocTracker::record_malloc();
  15.729 us [110348] |       } /* os::malloc */
  18.398 us [110348] |     } /* ResourceObj::operator new */
   1.124 us [110348] |     TimeStamp::update_to();
   2.608 us [110348] |     Arguments::process_sun_java_launcher_properties();
            [110348] |     os::init() {
   0.549 us [110348] |       ThreadCritical::initialize();
 113.820 us [110348] |       os::Linux::clock_init();
 147.623 us [110348] |     } /* os::init */
            [110348] |     Arguments::init_system_properties() {
            [110348] |       CHeapObj::operator new() {
```

### CODE REVIEW

File: libmcount/dynamic.c [hunk 1]
```diff
@@ -73,6 +74,39 @@ static struct mcount_orig_insn *lookup_code(struct rb_root *root,
...
 }
 
+static inline bool frozen(struct code_page *cp)
+{
+	return cp->frozen;
+}
+
+static inline int release(struct code_page *cp)
+{
+	int res = 0;
```
Comment (namhyung) [inline libmcount/dynamic.c]: no need to initialize.

File: libmcount/dynamic.c [hunk 2]
```diff
@@ -73,6 +74,39 @@ static struct mcount_orig_insn *lookup_code(struct rb_root *root,
...
+
+	res = mprotect(cp->page, CODE_CHUNK, PROT_READ|PROT_EXEC|PROT_WRITE);
+	if (res == 0)
+		cp->frozen = false;
+	else
+		pr_dbg("release failed as unexpected.");
```
Comment (namhyung) [inline libmcount/dynamic.c]: I think it's dangerous to make the code page writable after it contains executable code actually.  I think it's better to allocate a new separate codepage for dlopen().

**Comment by namhyung** [comment]
This should fix #817.

**Comment by ParkHanbum** [comment]
@namhyung could you review this PR again, please?

File: libmcount/dynamic.c [hunk 3]
```diff
@@ -514,14 +514,33 @@ static void finish_dynamic_update()
...
 
 	*rem = (100 * n - quot * total) * 100 / total;
 	return quot;
 }
 
+void print_dynamic_status(const char *filename)
+{
+	unsigned int r = 0, q = 0;
+
+	if (stats.total) {
```
Comment (namhyung) [inline libmcount/dynamic.c]: Actually it's not a mistake and I thought it's not necessary.

File: libmcount/dynamic.c [hunk 4]
```diff
@@ -73,6 +74,26 @@ static struct mcount_orig_insn *lookup_code(struct rb_root *root,
...
 
+static inline bool frozen(struct code_page *cp)
+{
+	return cp->frozen;
+}
+
+static struct code_page *alloc_codepage()
+{
+	struct code_page *cp = NULL;
+	cp = xmalloc(sizeof(*cp));
```
Comment (namhyung) [inline libmcount/dynamic.c]: You'd better using `xzalloc()` to zero-initialize.

File: libmcount/dynamic.c [hunk 5]
```diff
@@ -73,6 +74,26 @@ static struct mcount_orig_insn *lookup_code(struct rb_root *root,
 	return iter;
 }
 
+static inline bool frozen(struct code_page *cp)
+{
+	return cp->frozen;
+}
+
+static struct code_page *alloc_codepage()
```
Comment (namhyung) [inline libmcount/dynamic.c]: Please add `void` in the argument, it's a standard in C.

File: libmcount/dynamic.c (hunk 3 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c]: if user use verbose option that mean user expect statistic also. aren't you?

File: libmcount/dynamic.c (hunk 4 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c]: thanks for advice, I'll change it as follow as your comment.

File: libmcount/dynamic.c (hunk 5 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c]: thanks for advice, I'll change it as follow as your comment.

**Comment by ParkHanbum** [comment]
@namhyung sorry for that, I did not understand what you say, exactly, before. did you want function that support instrumenting to library with alias name in this PR? or separate it into another PR?

**Comment by namhyung** [comment]
> did you want function that support instrumenting to library with alias name in this PR? or separate it into another PR?

In a different PR please.

File: libmcount/dynamic.c (hunk 3 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c]: @namhyung I think all your point has been applied except this one. how do you think so?

Comment (namhyung) [reply libmcount/dynamic.c]: I still see many review comments above not answered yet.

File: libmcount/dynamic.c [hunk 6]
```diff
@@ -211,6 +211,43 @@ struct mcount_dynamic_info * create_mdi(unsigned long baseaddr,
 	return mdi;
 }
 
+
+struct module_list {
+	char** module_names;
+	uint32_t count;
+};
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c]: done!

File: libmcount/dynamic.c [hunk 7]
```diff
@@ -211,6 +211,43 @@ struct mcount_dynamic_info * create_mdi(unsigned long baseaddr,
...
+
+struct module_list module_list;
+
+char* find_module_list(char *modname)
+{
+	for (unsigned i = 0; i < module_list.count; i++) {
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c]: it was changed to use list. this code does not exist.

File: libmcount/dynamic.c [hunk 8]
```diff
@@ -607,3 +607,69 @@ bool mcount_add_badsym(struct mcount_dynamic_info *mdi, unsigned long callsite,
...
+	unsigned long vaddr = 0, text_addr = 0;
+	unsigned int text_size = 0, i;
+	bool text_section_found = false;
+
+	char* modname = basename(mod_realpath);
+
+	if (!find_module_list(modname))
+		return;
+
+	char *libname = strdup(modname);
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c]: done!

File: libmcount/dynamic.c [hunk 9]
```diff
@@ -607,3 +607,69 @@ bool mcount_add_badsym(struct mcount_dynamic_info *mdi, unsigned long callsite,
...
+		}
+	}
+
+	map->end += map->start;
+	text_addr += map->start;
+	map->mod = load_module_symtab(symtabs, mod_realpath);
+	maps_add_map(symtabs->maps, map);
+	mdi = create_mdi(map->start, text_addr, text_size);
+
+	if (map && map->mod) {
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c]: code changed as  your advice.

File: libmcount/internal.h [hunk 10]
```diff
@@ -495,10 +496,13 @@ void mcount_list_events(void);
 int mcount_arch_enable_event(struct mcount_event_info *mei);
 
 void mcount_hook_functions(void);
+void mcount_handle_dlopen(struct symtabs *, struct dl_phdr_info *,
+			  char *, size_t);
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/internal.h]: code changed as  your advice.

File: libmcount/dynamic.c [hunk 11]
```diff
@@ -519,14 +519,36 @@ static void finish_dynamic_update()
...
 
+void print_dynamic_status(const char *filename)
+{
+	unsigned int r = 0, q = 0;
+
+	if (stats.total) {
+		unsigned int success = stats.total - stats.failed - stats.skipped;
+
+		pr_dbg("dynamic patch stats for '%s'\n", basename(filename));
+		pr_dbg("   total: %8d\n", stats.total);
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c]: code changed as  your advice.

File: libmcount/dynamic.c [hunk 12]
```diff
@@ -519,14 +519,36 @@ static void finish_dynamic_update()
 }
 
 /* do not use floating-point in libmcount */
-static int calc_percent(int n, int total, int *rem)
+static int calc_percent(unsigned int n, unsigned int total, unsigned int *rem)
 {
+	if (n == 0)
+		return 100;
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c]: it was changed to use list. this code does not exist.

File: libmcount/dynamic.c (hunk 1 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c]: it was changed to use list. this code does not exist.

File: libmcount/dynamic.c (hunk 2 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c]: code changed as your advice.

File: libmcount/dynamic.c [hunk 13]
```diff
@@ -454,22 +457,21 @@ static int do_dynamic_update(struct symtabs *symtabs, char *patch_funcs,
 	return 0;
 }
 
-static void finish_dynamic_update(struct mcount_disasm_engine *disasm)
+static void finish_dynamic_update()
```
Comment (honggyukim) [inline libmcount/dynamic.c]: It's better to add explicit `void` for the function parameter.

File: libmcount/dynamic.c [hunk 14]
```diff
@@ -544,3 +637,71 @@ bool mcount_add_badsym(struct mcount_dynamic_info *mdi, unsigned long callsite,
...
+	unsigned int text_size = 0, i;
+	bool text_section_found = false;
+
+	char* modname = basename(mod_realpath);
+
+	if (!find_module_list(modname))
+		return;
+
+	char *libname = xstrdup(modname);
+	int namelen = ALIGN(strlen(libname) + 1, 4);
```
Comment (honggyukim) [inline libmcount/dynamic.c]: move these declarations with other declarations.

**Comment by honggyukim** [comment]
@ParkHanbum In addition, could you please write a PR as small as possible? It's very difficult to review such huge changes without spending a lot of time. That's the reason why this PR is pending for a long time.

I would like to ask you to send a preparation PR first so that it can be merged without spending a lot of effort. For example, I think the following commits can be done in a separate PR before this.
- 0e86bac dynaimc: add function print_dynamic_status and fixed one mistake.
- 91e3abb dynamic: disasm moved, min_size and disasm change scope to global
- 765d724 dynamic: function create_mdi added
- 7e96c30 utils: function maps_add_map added

If those commits are merged before this work, then the remaining review will be a lot easier and will not require many things to be changed again and again.

I believe that the core patch in this PR is only the followings.
- 51faf06 dynamic: add 'frozen' field to struct code_page
- 3a59fa3 dynamic: add mcount_handle_dlopen to handle dlopen

It will be a lot better to focus on reviewing only a few patches above.

File: utils/symbol.c [hunk 15]
```diff
@@ -1436,6 +1436,15 @@ struct sym * find_symname(struct symtab *symtab, const char *name)
 	return NULL;
 }
 
+void maps_add_map(struct uftrace_mmap *maps, struct uftrace_mmap *map)
+{
+	while(maps->next != NULL) {
+		maps = maps->next;
+	}
```
[the remark this answers is not in the document]
Comment (namhyung) [reply utils/symbol.c]: I personally prefer having an explicit NULL check.

File: libmcount/dynamic.c [hunk 16]
```diff
@@ -67,6 +74,26 @@ static struct mcount_orig_insn *lookup_code(struct rb_root *root,
 	return iter;
 }
 
+static inline bool frozen(struct code_page *cp)
+{
+	return cp->frozen;
+}
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c:82]: has been removed.

File: utils/symbol.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply utils/symbol.c:1443]: it hasn't changed yet. still you want to change it?

File: libmcount/dynamic.c [hunk 17]
```diff
@@ -211,6 +211,37 @@ struct mcount_dynamic_info * create_mdi(unsigned long baseaddr,
...
+	struct list_head	list;
+	char*			module_name;
+};
+
+static LIST_HEAD(module_list);
+
+struct module_list *find_module_list(char *modname)
+{
+	struct module_list *module;
+	list_for_each_entry(module, &module_list, list) {
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c:284]: sure, I missed it.

File: libmcount/internal.h [hunk 18]
```diff
@@ -495,10 +496,14 @@ void mcount_list_events(void);
...
+void mcount_handle_dlopen(struct symtabs *symtabs,
+			  struct dl_phdr_info *info, char *mod_realpath,
+			  size_t size);
 
 int read_pmu_event(struct mcount_thread_data *mtdp, enum uftrace_event_id id,
 		   void *buf);
 void release_pmu_event(struct mcount_thread_data *mtdp, enum uftrace_event_id id);
 void finish_pmu_event(struct mcount_thread_data *mtdp);
 
+
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/internal.h]: it's my bad

File: libmcount/dynamic.c [hunk 19]
```diff
@@ -73,6 +74,26 @@ static struct mcount_orig_insn *lookup_code(struct rb_root *root,
...
 
+static inline bool frozen(struct code_page *cp)
+{
+	return cp->frozen;
+}
+
+static struct code_page *alloc_codepage(void)
+{
+	struct code_page *cp = NULL;
+	cp = xzalloc(sizeof(*cp));
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c:88]: sure, I missed it.

File: libmcount/dynamic.c [hunk 20]
```diff
@@ -94,17 +115,25 @@ struct mcount_orig_insn *mcount_save_code(struct mcount_disasm_info *info,
...
 	}
 
 	orig = lookup_code(&code_tree, info->addr, true);
+
+	/*
+	 * if dynamic patch has been processed before, cp be frozen by
+	 * calling freeze_code. so, when reaching here from the
+	 * mcount_handle_dlopen, cp unwriteable.
+	 */
+	if (frozen(cp)) {
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c]: I changed it like you said.

File: libmcount/dynamic.c [hunk 21]
```diff
@@ -360,142 +507,88 @@ static int do_dynamic_update(struct symtabs *symtabs, char *patch_funcs,
...
 
-		mcount_arch_dynamic_recover(mdi, disasm);
+		mcount_arch_dynamic_recover(mdi, &disasm);
 		mcount_cleanup_trampoline(mdi);
 		free(mdi);
 
 		mdi = tmp;
 	}
 
-	mcount_disasm_finish(disasm);
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c:491]: uftrace supports tracing to dynamically loaded modules. so, changed the scope of disasm to global. do you think it is better to reallocating disasm each time?

File: libmcount/dynamic.c (hunk 13 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c]: I missed it. thanks.

File: libmcount/dynamic.c (hunk 14 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c]: was reflected what you said.

**Comment by ParkHanbum** [comment]
this PR has been updated. sorry for late.

**Comment by namhyung** [comment]
Hi Hanbum, thanks for the update!  I'll take a look later..

File: libmcount/dynamic.c [hunk 22]
```diff
@@ -643,6 +643,74 @@ bool mcount_add_badsym(struct mcount_dynamic_info *mdi, unsigned long callsite,
...
+	unsigned int text_size = 0, namelen, i;
+	bool text_section_found = false;
+
+	modname = basename(mod_realpath);
+
+	if (!find_module_list(modname))
+		return;
+
+	libname = xstrdup(modname);
+	namelen = ALIGN(strlen(libname) + 1, 4);
```
Comment (namhyung) [inline libmcount/dynamic.c:688]: Now I think we don't need this `ALIGN(len, 4)` part.

File: libmcount/dynamic.c [hunk 23]
```diff
@@ -643,6 +643,74 @@ bool mcount_add_badsym(struct mcount_dynamic_info *mdi, unsigned long callsite,
...
+	unsigned long vaddr = 0, text_addr = 0;
+	unsigned int text_size = 0, namelen, i;
+	bool text_section_found = false;
+
+	modname = basename(mod_realpath);
+
+	if (!find_module_list(modname))
+		return;
+
+	libname = xstrdup(modname);
```
Comment (namhyung) [inline libmcount/dynamic.c:687]: I don't think we need this since libname is a part of a map.

File: libmcount/dynamic.c [hunk 24]
```diff
@@ -643,6 +643,74 @@ bool mcount_add_badsym(struct mcount_dynamic_info *mdi, unsigned long callsite,
...
+	namelen = ALIGN(strlen(libname) + 1, 4);
+	map = xzalloc(sizeof(*map) + namelen);
+	map->start = info->dlpi_addr;
+	map->len = namelen;
+	mcount_memcpy1(map->libname, libname, namelen);
+	map->libname[namelen] = '\0';
+
+	map->end = 0;
```
Comment (namhyung) [inline libmcount/dynamic.c:695]: This is not needed since map is allocated by xzalloc() which sets all memory to zero (that's why it has the 'z' prefix).

File: libmcount/dynamic.c [hunk 25]
```diff
@@ -643,6 +643,74 @@ bool mcount_add_badsym(struct mcount_dynamic_info *mdi, unsigned long callsite,
...
+	maps_add_map(symtabs->maps, map);
+	mdi = create_mdi(map->start, text_addr, text_size);
+
+	if (map->mod) {
+		mdi->map = map;
+		mcount_arch_find_module(mdi, &map->mod->symtab);
+
+		mdi->next = mdinfo;
+		mdinfo = mdi;
+		mdi = setup_trampoline(map);
```
Comment (namhyung) [inline libmcount/dynamic.c:727]: It seems better to call `mcount_setup_trampoline()` directly.

File: libmcount/dynamic.c [hunk 26]
```diff
@@ -643,6 +643,74 @@ bool mcount_add_badsym(struct mcount_dynamic_info *mdi, unsigned long callsite,
...
+		if (mdi == NULL)
+			return;
+
+	}
+	else {
+		free(mdi);
+		return;
+	}
```
Comment (namhyung) [inline libmcount/dynamic.c:735]: Maybe we can file earlier than creating `mdi` in this case?

File: libmcount/dynamic.c [hunk 27]
```diff
@@ -101,20 +118,27 @@ struct mcount_orig_insn *mcount_save_code(struct mcount_disasm_info *info,
...
+	/*
+	 * if dynamic patch has been processed before, cp be frozen by
+	 * calling freeze_code. so, when reaching here from the
+	 * mcount_handle_dlopen, cp unwriteable.
+	 */
+	if (cp->frozen) {
+		/* [Caution]
+		 * even if a little memory loss occurs, it can be dangerous
+		 * that to re-assigned write and execute permission to exist
+		 * codepage, so be sure to allocate new memory.
```
Comment (namhyung) [inline libmcount/dynamic.c:137]: Agreed!

File: tests/t248_dynamic_dlopen.py [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+        return TestBase.build_libmain(self, name, 's-dlopen.c', ['libdl.so'],
+                                      cflags, ldflags)
+
+
+    def runcmd(self):
+        uftrace = TestBase.uftrace_cmd
+        options = TestBase.default_opt
+        options += ' -P. -P.@libfoo.so -P.@libabc_test_lib.so '
+        program = 't-' + self.name
+        return '%s %s %s' % (uftrace, options, program)
```
Comment (namhyung) [inline tests/t248_dynamic_dlopen.py:42]: Please update the test case to follow the new style.  You can simply set self.option IMO.

File: libmcount/dynamic.c [hunk 29]
```diff
@@ -196,7 +213,8 @@ static int find_dynamic_module(struct dl_phdr_info *info, size_t sz, void *data)
 	struct symtabs *symtabs = fmd->symtabs;
 	struct uftrace_mmap *map;
 	bool base_addr_set = false;
-	unsigned i;
+	unsigned long base_addr = 0, text_addr = 0;
+	unsigned int text_size = 0, i;
 
 	mdi = xzalloc(sizeof(*mdi));
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c:317]: it's my mistake sorry about that.

File: libmcount/dynamic.c [hunk 30]
```diff
@@ -357,22 +359,81 @@ static bool match_pattern_list(struct list_head *patterns,
...
+
+	if (!found)
+		stats.nomatch++;
+
+}
+
+static void mcount_patch_funcs(struct mcount_dynamic_info *mdi,
+			       struct uftrace_mmap *map)
+{
+	patch_func_matched(&patterns, mdi, map);
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c:484]: I think I was tried to keep the mcount prefixed format. not it removed.

File: libmcount/dynamic.c (hunk 23 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c:687]: I cannot understand it. could you elaborate more?

File: libmcount/dynamic.c (hunk 22 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c:688]: then ok, it was removed.

File: libmcount/dynamic.c (hunk 24 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c:695]: I did not know it! it has be changed.

File: libmcount/dynamic.c (hunk 25 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c:727]: thanks for advise!

File: libmcount/dynamic.c (hunk 26 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c:735]: you right, and I tihnk it is not need to check null for mdi.

File: libmcount/dynamic.c [hunk 31]
```diff
@@ -56,6 +56,9 @@ static struct Hashmap *code_hmap;
 /* at least, function size must be greater than this */
 static uint64_t min_size = 0;
 
+/* disassembly engine for dynamic code patch */
+static struct mcount_disasm_engine disasm;
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c:61]: commit message changed to filescope from globalscope.

File: libmcount/dynamic.c (hunk 27 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c:137]: actually, it was you advise.

File: tests/t248_dynamic_dlopen.py (hunk 28 above)
Comment (ParkHanbum) [reply tests/t248_dynamic_dlopen.py]: I see. it was changed.

File: libmcount/dynamic.c [hunk 32]
```diff
@@ -244,6 +244,38 @@ struct mcount_dynamic_info * create_mdi(unsigned long baseaddr,
...
+	list_for_each_entry(module, &module_list, list) {
+		if (!strncmp(module->module_name, modname, strlen(modname)))
+			return module;
+	}
+
+	return NULL;
+}
+
+
+void add_module_list(char *modname)
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply libmcount/dynamic.c]: add_module_list changed to static. any others?

**Comment by ParkHanbum** [comment]
I added a function to release resouces which allocated dynamically and globaly.

File: libmcount/dynamic.c (hunk 29 above)
[the remark this answers is not in the document]
Comment (namhyung) [reply libmcount/dynamic.c:317]: You didn't change it..

File: libmcount/dynamic.c [hunk 33]
```diff
@@ -357,22 +359,75 @@ static bool match_pattern_list(struct list_head *patterns,
...
+			case INSTRUMENT_SUCCESS:
+			default:
+				break;
+		}
+		stats.total++;
+	}
+
+	if (!found)
+		stats.nomatch++;
+
```
Comment (namhyung) [inline libmcount/dynamic.c:478]: Please remove this unnecessary blank line.

File: libmcount/dynamic.c [hunk 34]
```diff
@@ -244,6 +244,38 @@ struct mcount_dynamic_info * create_mdi(unsigned long baseaddr,
 	return mdi;
 }
 
+struct module_list {
+	struct list_head	list;
+	char*			module_name;
```
Comment (namhyung) [inline libmcount/dynamic.c:275]: Rename it to 'name' so that we can use 'module->name' later.

File: libmcount/dynamic.c (hunk 23 above)
Comment (namhyung) [reply libmcount/dynamic.c:696]: This libname will be leaked (and not needed) since we allocate space for name + map below.  Just remove this line..

File: libmcount/dynamic.c [hunk 35]
```diff
@@ -637,6 +637,66 @@ bool mcount_add_badsym(struct mcount_dynamic_info *mdi, unsigned long callsite,
...
+
+	if (!find_module_list(modname))
+		return;
+
+	libname = xstrdup(modname);
+	namelen = strlen(libname) + 1;
+	map = xzalloc(sizeof(*map) + namelen);
+	map->start = info->dlpi_addr;
+	map->len = namelen;
+	mcount_memcpy1(map->libname, libname, namelen);
```
Comment (namhyung) [inline libmcount/dynamic.c:701]: ... and use modname instead.

File: libmcount/dynamic.c (hunk 26 above)
Comment (namhyung) [reply libmcount/dynamic.c:736]: Did you change?  I mean we need to update `mdinfo` only if `mcount_setup_trampoline()` is succeeded.

File: libmcount/dynamic.c (hunk 33 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c:478]: I swear to God, I modified it before... At least it was fixed now.

File: libmcount/dynamic.c (hunk 34 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c:275]: it changed!

File: libmcount/dynamic.c (hunk 35 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c:701]: it changed.

File: libmcount/dynamic.c (hunk 26 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c:734]: because mdi could not be null, removed null check logic.

File: libmcount/dynamic.c (hunk 24 above)
Comment (namhyung) [reply libmcount/dynamic.c:700]: ??

File: libmcount/dynamic.c (hunk 26 above)
Comment (namhyung) [reply libmcount/dynamic.c:734]: I didn't talk about the NULL check, I'm saying about `mcount_setup_trampoline()`.  If it fails, we probably don't want to have the `mdi` so better to free it.

File: libmcount/dynamic.c [hunk 36]
```diff
@@ -244,6 +244,38 @@ struct mcount_dynamic_info * create_mdi(unsigned long baseaddr,
 	return mdi;
 }
 
+struct module_list {
+	struct list_head	list;
+	char*			name;
+};
+
+static LIST_HEAD(module_list);
```
Comment (namhyung) [inline libmcount/dynamic.c:278]: Why is this needed?  Who frees it?

File: libmcount/dynamic.c (hunk 26 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c:745]: I now understand what you say. it changed.

File: libmcount/dynamic.c (hunk 24 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c]: I didn't know your saying has include code that giving 0 to the map->end. it changed.

File: libmcount/dynamic.c (hunk 36 above)
Comment (ParkHanbum) [reply libmcount/dynamic.c:276]: First, I updated the code and added the code to free the module_list. and It is needed to keep a list of modules that have not yet been instrumented.

File: libmcount/dynamic.c (hunk 26 above)
Comment (namhyung) [reply libmcount/dynamic.c:743]: I don't think you did it correctly. Also you freed the `mdi` twice.  What I want is

```c
   mdi = create_mdi(...);
   if (mcount_setup_trampoline(mdi) < 0) {
      free(mdi);
      return;
   }
   mdi->next = mdi;
   mdinfo = mdi;
```

File: libmcount/dynamic.c (hunk 36 above)
Comment (namhyung) [reply libmcount/dynamic.c:276]: So you did it only for dlopen, right?  I'm not sure this is necessary as we want to patch functions in dlopen library always..  What's the actual case (or problematic case) do you see?

**Comment by kbumsik** [comment]
I just needed to track dlopen'ed functions and tried this branch. It works pretty well. Thanks for your effort!

**Comment by namhyung** [comment]
@kbumsik thanks for the testing!

@ParkHanbum I think the functionality is almost complete, I'll take a look and fix misc issues and then will merge the change. Thanks!

**Comment by ParkHanbum** [comment]
@namhyung Let me know if you have any request to do.

**Comment by namhyung** [comment]
@ParkHanbum Hi, I think you didn't address my feedback last time.  Please check it thoroughly from the first commit.

**Comment by ParkHanbum** [comment]
@namhyung I'm sorry about that. I'll review it again this weekend.

**Comment by namhyung** [comment]
No worries and thanks for your work.

File: libmcount/dynamic.c (hunk 29 above)
[the remark this answers is not in the document]
Comment (namhyung) [reply libmcount/dynamic.c:333]: @ParkHanbum please change this

File: libmcount/dynamic.c [hunk 37]
```diff
@@ -470,7 +519,8 @@ static int do_dynamic_update(struct symtabs *symtabs, char *patch_funcs,
 		else
 			init_filter_pattern(PATT_GLOB, &pl->patt, "*");
 
-		list_add(&pl->list, &patterns);
+		// add pl to all modules
+		// list_add(&pl->list, &patterns);
```
Comment (namhyung) [inline libmcount/dynamic.c:573]: What is this?

File: libmcount/dynamic.c [hunk 38]
```diff
@@ -58,6 +58,22 @@ static uint64_t min_size = 0;
 /* disassembly engine for dynamic code patch */
 static struct mcount_disasm_engine disasm;
 
+static struct list_head *instrument_ml;
```
Comment (namhyung) [inline libmcount/dynamic.c:62]: Where is it initialized?

File: libmcount/dynamic.c [hunk 39]
```diff
@@ -58,6 +58,22 @@ static uint64_t min_size = 0;
...
+	struct uftrace_pattern patt;
+	char *module;
+	bool positive;
+};
+
+struct module_list {
+	struct list_head list;
+	struct list_head patt_list;
+	char *modname;
+};
```
Comment (namhyung) [inline libmcount/dynamic.c:75]: I don't know why you want to keep the pattern list in the module_list entry.

**Review by namhyung — COMMENTED** [review]
Please notice me (via a comment here or a chat) if you update some code and want to get review from me.  Btw did you test this code?  And I see some review comments were still not addressed..

**Comment by namhyung** [comment]
@ParkHanbum Hi, I've uploaded this code with some modification to the [review/dynamic-dlopen-v1](https://github.com/namhyung/uftrace/compare/review/dynamic-dlopen-v1) branch.  Please take a look.

**Comment by namhyung** [comment]
I'll merge the above change.  @ParkHanbum please let me know if you have any concerns.

**Comment by namhyung** [comment]
Merged: fe09c869
